File: spojPractice/spoj3375.java
import java.io.*;
class spoj3375
{
	public static void main(String args[])throws IOException
	{
		BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
		spoj3375 obj=new spoj3375();
		int sc=1;
		int t=Integer.parseInt(br.readLine());
		for(int q=1;q<=t;q++)
		{
			
			//int b=Integer.parseInt(br.readLine());
			//int fr=Integer.parseInt(br.readLine());
			String lines=br.readLine();
			String[] strs=lines.trim().split("\\s+");
			int w[]=new int[2];
			for(int i=0;i<2;i++)
			{
				w[i]=Integer.parseInt(strs[i]);
			}
			int b=w[0];
			int fr=w[1];
			String lines1=br.readLine();
			String[] strs1=lines1.trim().split("\\s+");
			int a[]=new int[fr];
			for(int i=0;i<fr;i++)
			{
				a[i]=Integer.parseInt(strs1[i]);
			}
			a=obj.mergeSort(a);
			int f=0;
			int sum=0;
			int ans=0;
			for(int i=fr-1;i>=0;i--)
			{
				sum+=a[i];
				if(sum>=b)
				{
					ans=fr-i;
					f=1;
					break;
				}
			}
			System.out.println("Scenario #"+sc+":");
			if(f==1)
				System.out.println(ans);
			else
				System.out.println("impossible");
			sc++;
			System.out.println();
		}
	}
	public int[] mergeSort(int [] a)
    {
        if(a.length<=1)
        {
            return a;
        }
        int firstLength=a.length/2;
        int secondLength=a.length-firstLength;
        int first[]=new int[firstLength];
        int second[]=new int[secondLength];
        System.arraycopy(a,0,first,0,firstLength);
        System.arraycopy(a,firstLength,second,0,secondLength);
        first=mergeSort(first);
        second=mergeSort(second);
        merge(first,second,a);
        return a;
    } 

    public void merge(int [] first,int [] second,int [] a)
    {
        int iFirst=0;
        int iSecond=0;
        int j=0;
        while(iFirst<first.length&&iSecond<second.length)
        {
            if(first[iFirst]<second[iSecond])
            {
                a[j]=first[iFirst];
                iFirst++;
            }
            else
            {
                a[j]=second[iSecond];
                iSecond++;
            }
            j++;
        }
        System.arraycopy(first,iFirst,a,j,first.length-iFirst);
        System.arraycopy(second,iSecond,a,j,second.length-iSecond);
    }
}

File: spojPractice/spoj3375.cpp
#include <iostream>
using namespace std;
void merge(int *a, int low, int high, int mid)
{
    int i, j, k, c[50];
    i = low;
    k = low;
    j = mid + 1;
    while (i <= mid && j <= high)
    {
        if (a[i] < a[j])
        {
            c[k] = a[i];
            k++;
            i++;
        }
        else
        {
            c[k] = a[j];
            k++;
            j++;
        }
    }
    while (i <= mid)
    {
        c[k] = a[i];
        k++;
        i++;
    }
    while (j <= high)
    {
        c[k] = a[j];
        k++;
        j++;
    }
    for (i = low; i < k; i++)
    {
        a[i] = c[i];
    }
}
void mergesort(int *a, int low, int high)
{
    int mid;
    if (low < high)
    {
        mid=(low+high)/2;
        mergesort(a,low,mid);
        mergesort(a,mid+1,high);
        merge(a,low,high,mid);
    }
    return;
}
int main()
{
	int t;
	int q=1;
	cin>>t;
	while(t--)
	{
		int br;
		int fr;
		cin>>br>>fr;

		int m[100000];
		//m=new int[fr];
		for(int i=0;i<fr;i++)
			cin>>m[i];
		mergesort(m,0,fr-1);
		int sum=0;
		int ans=0;
		//for(int i=0;i<fr;i++)
		//	cout<<m[i]<<" ";
		//cout<<endl;
		//cout<<br<<"br"<<endl;
		for(int i=fr-1;i>0;i--)
		{
			sum+=m[i];
			//cout<<sum<<" ";
			if(sum>=br)
			{
				ans=fr-i;
				break;
			}
		}
		if(ans!=0)
		{
			cout<<"Scenario #"<<q<<":\n";
			cout<<ans<<endl;
			q++;
		}
		else
		{
			cout<<"Scenario #"<<q<<":\n";
			cout<<"impossible"<<endl;
			q++;
		}
		cout<<endl;
	}
}